rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java

```diff
@@ -384,37 +385,74 @@ else if (isCached.get(idx).getLocation() == Location.CACHED) {
             else {
                 if (isCached.get(idx).getLocation() == Location.NON_LOCAL) {
```

Comment:
Is it possible to split outall Non-local code to a new class ? It feels like this code needs an interface and derived classes. However, I cannot be sure with the current state. Also as a novice in this code, I am unable to effectively review this part.